store/cachemulti/store.go

```diff
@@ -0,0 +1,133 @@
+package cachemulti
+
+import (
+	"io"
+
+	dbm "github.com/tendermint/tendermint/libs/db"
+
+	"github.com/cosmos/cosmos-sdk/store/cachekv"
+	"github.com/cosmos/cosmos-sdk/store/dbadapter"
+	"github.com/cosmos/cosmos-sdk/store/types"
+)
+
+//----------------------------------------
+// Store
+
+// Store holds many cache-wrapped stores.
+// Implements MultiStore.
+// NOTE: a Store (and MultiStores in general) should never expose the
+// keys for the substores.
+type Store struct {
+	db         types.CacheKVStore
+	stores     map[types.StoreKey]types.CacheWrap
+	keysByName map[string]types.StoreKey
+
+	traceWriter  io.Writer
+	traceContext types.TraceContext
+}
+
+var _ types.CacheMultiStore = Store{}
+
+func NewFromKVStore(store types.KVStore, stores map[types.StoreKey]types.CacheWrapper, keysByName map[string]types.StoreKey, traceWriter io.Writer, traceContext types.TraceContext) Store {
+	cms := Store{
+		db:           cachekv.NewStore(store),
+		stores:       make(map[types.StoreKey]types.CacheWrap, len(stores)),
+		keysByName:   keysByName,
+		traceWriter:  traceWriter,
+		traceContext: traceContext,
+	}
+
+	for key, store := range stores {
+		if cms.TracingEnabled() {
+			cms.stores[key] = store.CacheWrapWithTrace(cms.traceWriter, cms.traceContext)
+		} else {
+			cms.stores[key] = store.CacheWrap()
+		}
+	}
+
+	return cms
+}
+
+func NewStore(db dbm.DB, stores map[types.StoreKey]types.CacheWrapper, keysByName map[string]types.StoreKey, traceWriter io.Writer, traceContext types.TraceContext) Store {
+	return NewFromKVStore(dbadapter.Store{db}, stores, keysByName, traceWriter, traceContext)
+}
+
+func newCacheMultiStoreFromCMS(cms Store) Store {
+	stores := make(map[types.StoreKey]types.CacheWrapper)
+	for k, v := range cms.stores {
+		stores[k] = v
+	}
+	return NewFromKVStore(cms.db, stores, nil, cms.traceWriter, cms.traceContext)
+}
+
+// WithTracer sets the tracer for the MultiStore that the underlying
+// stores will utilize to trace operations. A MultiStore is returned.
+func (cms Store) WithTracer(w io.Writer) types.MultiStore {
+	cms.traceWriter = w
+	return cms
+}
+
+// WithTracingContext updates the tracing context for the MultiStore by merging
+// the given context with the existing context by key. Any existing keys will
+// be overwritten. It is implied that the caller should update the context when
+// necessary between tracing operations. It returns a modified MultiStore.
+func (cms Store) WithTracingContext(tc types.TraceContext) types.MultiStore {
```

Comment:
This function is confusing and should be reworked, ref https://github.com/cosmos/cosmos-sdk/issues/2824

Reply:
Will be handled in the third PR: https://github.com/cosmos/cosmos-sdk/issues/2986

Reply:
Will be handled in the third PR, for now changed the names to `Set*`